### PR TITLE
Add support for custom prediction function in colab and jupyter notebook modes.

### PR DESCRIPTION
**Pull request by tolga-b**
* Motivation for features / changes

Add custom prediction function support in colab and jupyter mode so it is easier to load non-TensorFlow models into WIT.

* Technical description of changes

WITConfigBuilder now has a "set_custom_predict_fn" method to pass a python function for prediction purposes. Parts of notebook backend and inference_utils are updated to use the custom function when provided.

* Screenshots of UI changes

No UI changes.

* Detailed steps to verify changes work correctly (as executed by you)

Ran WIT Toxicity Text Model Comparison and WIT Model Comparison colabs with custom_predict_fn to verify the tool is getting inferences correctly. 

* Alternate designs / implementations considered

### CODE REVIEW

**Comment by tolga-b** [comment]
@jameswex please review.

File: tensorboard/plugins/interactive_inference/README.md [hunk 1]
```diff
@@ -57,6 +57,11 @@ You can use the What-If Tool to analyze a classification or regression
 that takes TensorFlow Example or SequenceExample protos
 (data points) as inputs directly in a jupyter or colab notebook.
 
+You can also use What-If-Tool with your own custom prediction function that takes
+Tensorflow Example and produces predictions. In this mode, you can load any model
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/README.md]: I would say "takes TensorFlow examples and produces predictions" (that way it includes Examples and SequenceExamples)

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py [hunk 2]
```diff
@@ -181,14 +181,16 @@ class ServingBundle(object):
       Predict API.
     estimator: An estimator to use instead of calling an external model.
     feature_spec: A feature spec for use with the estimator.
+    custom_predict_fn: A custom prediction function
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/utils/inference_utils.py]: Add period

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py [hunk 3]
```diff
@@ -336,14 +340,18 @@ def set_estimator_and_feature_spec(self, estimator, feature_spec):
     Returns:
       self, in order to enabled method chaining.
     """
+    # If custom function is set, remove it before setting estimator
+    self.delete('custom_predict_fn')
+    self.delete('compare_custom_predict_fn')
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: i think this should jsut delete the custom_predict_fn, not the compare one

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py [hunk 4]
```diff
@@ -363,10 +371,77 @@ def set_compare_estimator_and_feature_spec(self, estimator, feature_spec):
     Returns:
       self, in order to enabled method chaining.
     """
+    # If custom function is set, remove it before setting estimator
+    self.delete('custom_predict_fn')
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: i think this should only delete the custom compare predict fn

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py [hunk 5]
```diff
@@ -363,10 +371,77 @@ def set_compare_estimator_and_feature_spec(self, estimator, feature_spec):
...
+    Args:
+      predict_fn: The custom python function which will be used for model
+      inference.
+
+    Returns:
+      self, in order to enabled method chaining.
+    """
+    # If estimator is set, remove it before setting predict_fn
+    self.delete('estimator_and_spec')
+    self.delete('compare_estimator_and_spec')
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: same comment

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py [hunk 6]
```diff
@@ -363,10 +371,77 @@ def set_compare_estimator_and_feature_spec(self, estimator, feature_spec):
...
+
+    Args:
+      predict_fn: The custom python function which will be used for model
+      inference.
+
+    Returns:
+      self, in order to enabled method chaining.
+    """
+    # If estimator is set, remove it before setting predict_fn
+    self.delete('estimator_and_spec')
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: same comment

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py (hunk 3 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: done, makes sense if you want to compare an estimator to a predict function.

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py (hunk 4 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: done

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py (hunk 5 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: done

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py (hunk 6 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: done

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py (hunk 2 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/utils/inference_utils.py]: done

File: tensorboard/plugins/interactive_inference/README.md (hunk 1 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/README.md]: done!

**Comment by tolga-b** [comment]
Thanks James, I added all your comments.

**Pull request merged**